index.js

```diff
@@ -15,52 +18,57 @@ const serializeError = FJS({
 })
 
 function rateLimitPlugin (fastify, opts, next) {
-  const cache = lru(opts.cache || 5000)
-  const max = opts.max || 1000
-  const whitelist = opts.whitelist || []
   const timeWindow = typeof opts.timeWindow === 'string'
     ? ms(opts.timeWindow)
     : typeof opts.timeWindow === 'number'
-    ? opts.timeWindow
-    : 1000 * 60
-  const after = ms(timeWindow, { long: true })
+      ? opts.timeWindow
+      : 1000 * 60
 
-  const interval = setInterval(cache.reset.bind(cache), timeWindow)
-  if (interval.unref) interval.unref()
+  const store = opts.store
+    ? new RedisStore(opts.store, timeWindow)
+    : new LocalStore(lru(opts.cache || 5000), timeWindow)
+
+  const max = opts.max || 1000
+  const whitelist = opts.whitelist || []
+  const after = ms(timeWindow, { long: true })
 
   fastify.addHook('onRequest', onRateLimit)
 
   function onRateLimit (req, res, next) {
     var ip = req.headers['X-Forwarded-For'] || req.connection.remoteAddress
-    var current = cache.get(ip) || 0
-
-    var limitReached = current >= max
-
-    if (whitelist.indexOf(ip) === -1) {
-      if (limitReached === false) current++
-      cache.set(ip, current)
-      res.setHeader('X-RateLimit-Limit', max)
-      res.setHeader('X-RateLimit-Remaining', max - current)
-    }
-
-    if (limitReached === false) {
+    if (whitelist.indexOf(ip) > -1) {
       next()
     } else {
-      res.setHeader('Content-Type', 'application/json')
-      res.setHeader('Retry-After', timeWindow)
-      res.statusCode = 429
-      res.end(serializeError({
-        statusCode: 429,
-        error: 'Too Many Requests',
-        message: `Rate limit exceeded, retry in ${after}`
-      }))
+      store.incr(ip, onIncr)
+    }
+
+    function onIncr (err, current) {
+      if (err) return next(err)
+
+      if (current <= max) {
+        res.setHeader('X-RateLimit-Limit', max)
+        res.setHeader('X-RateLimit-Remaining', max - current)
+        next()
+      } else {
+        res.writeHead(429, {
+          'X-RateLimit-Limit': max,
+          'X-RateLimit-Remaining': 0,
+          'Content-Type': 'application/json',
+          'Retry-After': timeWindow
+        })
+        res.end(serializeError({
+          statusCode: 429,
+          error: 'Too Many Requests',
+          message: `Rate limit exceeded, retry in ${after}`
+        }))
+      }
     }
   }
 
   next()
```

Comment:
Is there any chance that Redis might not be connected? Should we wait until it's ready?

Reply:
I assume that the users passes us a working instance of Redis.

Reply:
a little bit more of safety here would be nice.

Reply:
it's probably better to verify. We don't really want to start listening with a broken Redis. Or maybe we want, but that needs to be checked properly.

Reply:
even if it's handled automatically, you truly want your Redis connection to be there when your server starts listening.

Reply:
As far as I know [`ioredis`](https://github.com/luin/ioredis) does not have a way to know if the database is working or not.
We can listen for redis events, but I think it's an overkill solution.
Otherwise we can try a basic query, but I guess it's overkill as well.

I think that if you are using redis in your application you are also using [`fastify-redis`](https://github.com/fastify/fastify-redis) which should handle all this problems for you.

Reply:
I agree with @mcollina that Redis MUST be online and available. Note that this should support either redis or sentinel.

Reply:
I've added a `skipOnError` option, checkout the latest code :)